DAY 22 ADVANCED GROUPBY AND BUSINESS METRICS

In [2]:
import pandas as pd

In [4]:
df = pd.read_csv("Data_Co_Supply_Chain_Dataset.csv" ,
encoding="latin")
df.shape

(180519, 31)

In [5]:
df.columns

Index(['Benefit per order', 'Sales per customer', 'Delivery Status',
       'Late_delivery_risk', 'Category Name', 'Customer City',
       'Customer Country', 'Customer Fname', 'Customer Id', 'Customer Lname',
       'Customer Segment', 'Customer State', 'Department Name', 'Market',
       'Order Country', 'order date (DateOrders)', 'Order Id',
       'Order Item Discount', 'Order Item Discount Rate',
       'Order Item Product Price', 'Order Item Profit Ratio',
       'Order Item Quantity', 'Sales', 'Order Item Total',
       'Order Profit Per Order', 'Order Region', 'Order State', 'Order Status',
       'Product Name', 'shipping date (DateOrders)', 'Shipping Mode'],
      dtype='object')

ADVANCED REGION ANALYSIS

In [10]:
region_analysis = df.groupby("Order Region").agg(
     Total_sales = ("Sales" , "sum"),
     Total_profit = ("Order Profit Per Order" , "sum"),
     Average_sales = ("Sales" , "mean"),
     Average_profit = ("Order Profit Per Order" , "mean"),
     Total_orders = ("Order Id" , "nunique")
)
region_analysis.head()

,Total_sales,Total_profit,Average_sales,Average_profit,Total_orders
Order Region,,,,,
Canada,1.868610e+05,23900.709960,194.849889,24.922534,309
Caribbean,1.651019e+06,171825.640024,198.487537,20.657086,2806
Central Africa,3.272630e+05,33447.269960,195.147893,19.944705,556
Central America,5.665712e+06,616341.570651,199.912216,21.747347,9396
Central Asia,1.098399e+05,13045.280051,198.625556,23.590018,184


PROFIT MARGIN %

In [12]:
region_analysis["Profit_Margin_%"] = (
     region_analysis["Total_profit"]
     .div(region_analysis["Total_sales"])
     .mul(100)
)
region_analysis.sort_values(
     by="Profit_Margin_%" ,
     ascending=False
)

,Total_sales,Total_profit,Average_sales,Average_profit,Total_orders,Profit_Margin_%
Order Region,,,,,,
Southern Africa,2.282516e+05,30826.050146,197.278820,26.643086,398,13.505295
Canada,1.868610e+05,23900.709960,194.849889,24.922534,309,12.790633
Central Asia,1.098399e+05,13045.280051,198.625556,23.590018,184,11.876628
East Africa,3.762349e+05,43167.729927,203.150593,23.308709,613,11.473611
East of USA,1.371112e+06,156263.300194,198.280837,22.597730,2323,11.396830
US Center,1.151356e+06,131094.229875,195.575976,22.268427,1935,11.386075
South America,2.960881e+06,335154.400817,198.251182,22.440870,4979,11.319413
Southern Europe,2.047919e+06,230829.229883,217.147579,24.475584,3543,11.271405
South of USA,7.857839e+05,88114.879898,194.260555,21.783654,1345,11.213627


SALES PER ORDER

In [14]:
region_analysis["Sales_Per_Order"] = (
     region_analysis["Total_sales"]
     .div(region_analysis["Total_orders"])
)
region_analysis.sort_values(
     by="Sales_Per_Order",
     ascending=False
)

,Total_sales,Total_profit,Average_sales,Average_profit,Total_orders,Profit_Margin_%,Sales_Per_Order
Order Region,,,,,,,
East Africa,3.762349e+05,43167.729927,203.150593,23.308709,613,11.473611,613.760029
Canada,1.868610e+05,23900.709960,194.849889,24.922534,309,12.790633,604.728296
Central America,5.665712e+06,616341.570651,199.912216,21.747347,9396,10.878448,602.991922
Eastern Europe,7.742666e+05,79717.049920,197.516981,20.335982,1292,10.295815,599.277527
Central Asia,1.098399e+05,13045.280051,198.625556,23.590018,184,11.876628,596.956153
North Africa,6.347522e+05,64599.860009,196.396108,19.987580,1064,10.177177,596.571638
West Africa,7.279512e+05,80030.270031,196.956492,21.653212,1223,10.993906,595.217657
US Center,1.151356e+06,131094.229875,195.575976,22.268427,1935,11.386075,595.015904
South America,2.960881e+06,335154.400817,198.251182,22.440870,4979,11.319413,594.673912


In [15]:
region_analysis["Sales_Per_Order"].idxmax()

'East Africa'

In [16]:
region_analysis["Sales_Per_Order"].idxmin()

'Southeast Asia'

## REGION + CUSTOMER SEGMENT

In [20]:
region_segment = df.groupby(
     ["Order Region" , "Customer Segment"]
)["Sales"].sum().reset_index()
region_segment.head(10)

,Order Region,Customer Segment,Sales
0,Canada,Consumer,9.542350e+04
1,Canada,Corporate,5.284418e+04
2,Canada,Home Office,3.859336e+04
3,Caribbean,Consumer,8.167988e+05
4,Caribbean,Corporate,5.285160e+05
5,Caribbean,Home Office,3.057045e+05
6,Central Africa,Consumer,1.609237e+05
7,Central Africa,Corporate,1.074103e+05
8,Central Africa,Home Office,5.892899e+04
9,Central America,Consumer,2.953943e+06


In [22]:
region_segment.sort_values(
     by="Sales" ,
     ascending=False
).head(10)

,Order Region,Customer Segment,Sales
66,Western Europe,Consumer,3.084309e+06
9,Central America,Consumer,2.953943e+06
67,Western Europe,Corporate,1.805624e+06
10,Central America,Corporate,1.716687e+06
36,South America,Consumer,1.518897e+06
30,Northern Europe,Consumer,1.110730e+06
51,Southern Europe,Consumer,1.065963e+06
33,Oceania,Consumer,1.051469e+06
68,Western Europe,Home Office,1.004448e+06
11,Central America,Home Office,9.950822e+05


WITHIN SEGMENT CONTRIBUTION

In [26]:
region_segment["Region_Total_Sales"] = (
     region_segment.groupby("Order Region")["Sales"]
     .transform("sum")
)
region_segment["Sales_Contribution_%"] = (
     region_segment["Sales"]
     .div(region_segment["Region_Total_Sales"])
     .mul(100)
)
region_segment.sort_values(
     by="Sales_Contribution_%" ,
     ascending=False
).head(10)

,Order Region,Customer Segment,Sales,Region_Total_Sales,Sales_Contribution_%
12,Central Asia,Consumer,6.457016e+04,1.098399e+05,58.785689
54,US Center,Consumer,6.233922e+05,1.151356e+06,54.144183
15,East Africa,Consumer,2.030225e+05,3.762349e+05,53.961628
42,South of USA,Consumer,4.213899e+05,7.857839e+05,53.626685
24,Eastern Europe,Consumer,4.105639e+05,7.742666e+05,53.026164
27,North Africa,Consumer,3.336980e+05,6.347522e+05,52.571381
60,West Asia,Consumer,6.154214e+05,1.174672e+06,52.390920
66,Western Europe,Consumer,3.084309e+06,5.894381e+06,52.326253
33,Oceania,Consumer,1.051469e+06,2.016654e+06,52.139295
9,Central America,Consumer,2.953943e+06,5.665712e+06,52.137182


DOMINANT SEGMENT IN EVERY REGION

In [28]:
dominat_segment = region_segment.loc[
     region_segment.groupby("Order Region")[
          "Sales_Contribution_%"
     ].idxmax()
]
dominat_segment.sort_values(
     by="Sales_Contribution_%" ,
     ascending = False
)

,Order Region,Customer Segment,Sales,Region_Total_Sales,Sales_Contribution_%
12,Central Asia,Consumer,6.457016e+04,1.098399e+05,58.785689
54,US Center,Consumer,6.233922e+05,1.151356e+06,54.144183
15,East Africa,Consumer,2.030225e+05,3.762349e+05,53.961628
42,South of USA,Consumer,4.213899e+05,7.857839e+05,53.626685
24,Eastern Europe,Consumer,4.105639e+05,7.742666e+05,53.026164
27,North Africa,Consumer,3.336980e+05,6.347522e+05,52.571381
60,West Asia,Consumer,6.154214e+05,1.174672e+06,52.390920
66,Western Europe,Consumer,3.084309e+06,5.894381e+06,52.326253
33,Oceania,Consumer,1.051469e+06,2.016654e+06,52.139295
9,Central America,Consumer,2.953943e+06,5.665712e+06,52.137182


## CUSTOMER VALUE ANALYSIS

In [32]:
#Customer-level aggregation
customer_analysis = df.groupby("Customer Id").agg(
     Total_sales = ("Sales" , "sum"),
     Total_orders = ("Order Id" , "nunique"),
     Average_order_value = ("Sales" , "mean")
     
)
customer_analysis.sort_values(
     by="Total_sales" ,
     ascending=False
).head(10)

,Total_sales,Total_orders,Average_order_value
Customer Id,,,
791,10524.170178,13,244.748144
9371,9299.030207,11,211.341596
8766,9296.140186,11,244.635268
1657,9223.710152,11,219.612146
2641,9130.920224,11,212.346982
1288,9019.110199,12,214.740719
3710,9019.100208,14,214.740481
4249,8918.850134,13,228.688465
5654,8904.950211,15,189.467026


Customer Sales Contribution

In [35]:
total_customer_sales = customer_analysis["Total_sales"].sum()

customer_analysis["Sales_Contribution_%"] = (
     customer_analysis["Total_sales"]
     .div(total_customer_sales).mul(100)
)
customer_analysis.sort_values(
     by="Sales_Contribution_%" ,
     ascending=False
).head(10)

,Total_sales,Total_orders,Average_order_value,Sales_Contribution_%
Customer Id,,,,
791,10524.170178,13,244.748144,0.028610
9371,9299.030207,11,211.341596,0.025280
8766,9296.140186,11,244.635268,0.025272
1657,9223.710152,11,219.612146,0.025075
2641,9130.920224,11,212.346982,0.024823
1288,9019.110199,12,214.740719,0.024519
3710,9019.100208,14,214.740481,0.024519
4249,8918.850134,13,228.688465,0.024246
5654,8904.950211,15,189.467026,0.024208


Customer Value: Order Value Comparison

In [37]:
customer_analysis["Average_order_value"].idxmax()

14057

In [38]:
customer_analysis.loc[
    customer_analysis["Average_order_value"].idxmax()
]

Total_sales             1500.000000
Total_orders               1.000000
Average_order_value     1500.000000
Sales_Contribution_%       0.004078
Name: 14057, dtype: float64

High-Value Customer Identification


In [40]:
high_value_customers = customer_analysis[
     (customer_analysis["Total_sales"] > 5000) &
     (customer_analysis["Total_orders"] >= 10) 
]
high_value_customers.sort_values(
     by="Total_sales" ,
     ascending = False
).head(10)

,Total_sales,Total_orders,Average_order_value,Sales_Contribution_%
Customer Id,,,,
791,10524.170178,13,244.748144,0.028610
9371,9299.030207,11,211.341596,0.025280
8766,9296.140186,11,244.635268,0.025272
1657,9223.710152,11,219.612146,0.025075
2641,9130.920224,11,212.346982,0.024823
1288,9019.110199,12,214.740719,0.024519
3710,9019.100208,14,214.740481,0.024519
4249,8918.850134,13,228.688465,0.024246
5654,8904.950211,15,189.467026,0.024208


## PRODUCT PERFORMANCE

In [41]:
product_analysis = df.groupby("Product Name").agg(
     Total_sales = ("Sales" , "sum"),
     Total_quantity = ("Order Item Quantity" , "sum"),
     Total_orders = ("Order Id" , "nunique")
)
product_analysis.sort_values(
     by="Total_sales" ,
     ascending=False
).head(10)

,Total_sales,Total_quantity,Total_orders
Product Name,,,
Field & Stream Sportsman 16 Gun Fire Safe,6.929654e+06,17325,15164
Perfect Fitness Perfect Rip Deck,4.421143e+06,73698,20359
Diamondback Women's Serene Classic Comfort Bi,4.118426e+06,13729,12299
Nike Men's Free 5.0+ Running Shoe,3.667633e+06,36680,11092
Nike Men's Dri-FIT Victory Golf Polo,3.147800e+06,62956,17869
Pelican Sunstream 100 Kayak,3.099845e+06,15500,13727
Nike Men's CJ Elite 2 TD Football Cleat,2.891758e+06,22246,18783
O'Brien Men's Neoprene Life Vest,2.888994e+06,57803,16623
Under Armour Girls' Toddler Spine Surge Runni,1.269083e+06,31735,9825


In [44]:
# sales contribution
total_product_sales = product_analysis["Total_sales"].sum()
product_analysis["Sales_Contribution_%"] = (
     product_analysis["Total_sales"]
     .div(total_product_sales)
     .mul(100)
)
product_analysis.sort_values(
     by="Sales_Contribution_%" ,
     ascending=False
).head(10)

,Total_sales,Total_quantity,Total_orders,Sales_Contribution_%
Product Name,,,,
Field & Stream Sportsman 16 Gun Fire Safe,6.929654e+06,17325,15164,18.838395
Perfect Fitness Perfect Rip Deck,4.421143e+06,73698,20359,12.018962
Diamondback Women's Serene Classic Comfort Bi,4.118426e+06,13729,12299,11.196018
Nike Men's Free 5.0+ Running Shoe,3.667633e+06,36680,11092,9.970530
Nike Men's Dri-FIT Victory Golf Polo,3.147800e+06,62956,17869,8.557354
Pelican Sunstream 100 Kayak,3.099845e+06,15500,13727,8.426988
Nike Men's CJ Elite 2 TD Football Cleat,2.891758e+06,22246,18783,7.861298
O'Brien Men's Neoprene Life Vest,2.888994e+06,57803,16623,7.853785
Under Armour Girls' Toddler Spine Surge Runni,1.269083e+06,31735,9825,3.450025


In [45]:
product_analysis["Sales_Contribution_%"].idxmax()

'Field & Stream Sportsman 16 Gun Fire Safe'

# Final Business Analysis

High Sales vs High Margin

In [47]:
(region_analysis["Profit_Margin_%"].idxmax() , 
region_analysis["Total_sales"].idxmax())

('Southern Africa', 'Western Europe')

High Value Customers

In [48]:
high_value_customers.shape[0]

227

Top Product

In [51]:
top_product = product_analysis["Total_sales"].idxmax()

product_analysis.loc[top_product]

Total_sales             6.929654e+06
Total_quantity          1.732500e+04
Total_orders            1.516400e+04
Sales_Contribution_%    1.883840e+01
Name: Field & Stream Sportsman 16 Gun Fire Safe, dtype: float64

## Final Business Insights — Day 22

1. Western Europe generates the highest total sales among all regions.

2. Southern Africa has the highest profit margin, showing stronger profitability efficiency despite not having the highest total sales.

3. Consumer is the dominant customer segment across the analyzed regions, with its contribution varying by region.

4. 227 customers meet the defined high-value criteria of generating more than 5,000 in sales and having at least 10 orders.

5. Field & Stream Sportsman 16 Gun Fire Safe is the highest-sales product, generating approximately 6.93M in sales and contributing approximately 18.84% of total product sales.

6. Central Africa has the highest late-delivery risk at approximately 57.96%, identifying a region that requires attention from a supply-chain perspective.

7. Regional performance differs depending on the metric: Western Europe leads total sales, while Southern Africa leads profit margin and East Africa leads sales per order.